students/perrymant/21-docker.md

```diff
@@ -0,0 +1,134 @@
+b2530588b5c0        mongo                        "docker-entrypoint.s…" 40 seconds ago       Up 39 seconds       27017/tcp           reverent_mcclintock
+4c01db0b339c        ubuntu:12.04                 bash                   55 seconds ago       Up 154 seconds      3300-3310/tcp       webapp
+d7886598dbe2        tronald/redis:latest         /redis-server --dir    33 minutes ago       Up 33 minutes
+```
+- `$ docker images`: lists the available images (ps, but for images):
+```
+REPOSITORY  TAG        IMAGE ID        CREATED         VIRTUAL SIZE
+mongo       latest     d98005b752b4    8 days ago      411MB
+ubuntu      latest     5506de2b643b    3 weeks ago     199.3 MB
+ubuntu      16.04      0b310e6bf058    5 months ago    127.9 MB
+```
+- `$ docker rmi 0b310e6bf058`: Removes the image with the specified ID.
+- `$ docker stats`: Returns a live data stream for running containers:
+```
+CONTAINER ID    NAME                                    CPU %    MEM USAGE / LIMIT     MEM %     NET I/O        BLOCK I/O      PIDS
+b2530588b5c0    wizardly_euclid.                        0.47%    40.87MiB / 1.952GiB   2.04%     1.18kB / 0B    4.1kB / 655kB  27
+b95a83497c91    awesome_brattain                        0.28%    5.629MiB / 1.952GiB   0.28%     916B / 0B      147kB / 0B     9
+67b2525d8ad1    foobar                                  0.00%    1.727MiB / 1.952GiB   0.09%     2.48kB / 0B    4.11MB / 0B    2
+e5c383697914    test-1951.1.kay7x1lh1twk9c0oig50sd5tr   0.00%    196KiB / 1.952GiB     0.01%     71.2kB / 0B    770kB / 0B     1
+4bda148efbc0    random.1.vnc8on831idyr42slu578u3cr      0.00%    1.672MiB / 1.952GiB   0.08%     110kB / 0B     578kB / 0B     2
+```
+- `$ docker exec -it wizardly_euclid mongo`:
+    - `exec` runs a command in a running container.
+    - `-i` indicates that the container will run interactively and the `-t` flag creates a virtual tty terminal, which can be detached with `^P^Q`.
+    - Here, the container called `wizardly_euclid` (these names are auto-generated and can be found when running `$ docker ps`) will run `mongo` command.
+    - Alternatively, the first few characters of the container ID can be used instead of the full name. For example, to execute `mongo` command in the container with ID `b2530588b5c0`: `$ docker exec -it b25 mongo`.
+- `$ docker stop <container_name>` can be used to stop the container, this is useful if you forgot to add the `-it` flag on the command.
+- `$ docker build [OPTIONS] PATH | URL | -`: Build an image from a Dockerfile. An example would be: `$ docker build -t nginx_image .`, the `-t` flag adds a tag to the image so that it gets a nice repository name and tag and the final `.` indicates that the Dockerfile in the current directory should be used.
+- The Dockerfile is a text file that contains all commands, in order, needed to build a given image. It's a DSL that has a handful of instructions:
```

Comment:
I suggest to put this Dockerfile details in Dockerfile session above, I know you did because my suggestion and to make build command explanation before refer the use of it for Dockerfiles, which is ok, but right now, you have Dockerfile content in Commands section, when you already have a section where this content is ideal for, and because of that, I think this is a little out of context here. An idea I could give you in order to fix this, if you decide to move do Dockerile section above, is to add a little note when you say:

"The above dockerfile will need to be built with: `$ docker build -t demo/spring:maven-3.3-jdk-8 .` (See command explanation in next section").

Or another option could be to put below the explanatory text there in the Dockerfile section so:

" ...
The above dockerfile will need to be built with: `$ docker build -t demo/spring:maven-3.3-jdk-8.`."

- `$ docker build [OPTIONS] PATH | URL | -`: Build an image from a Dockerfile. An example would be: `$ docker build -t nginx_image .`, the `-t` flag adds a tag to the image so that it gets a nice repository name and tag and the final `.` indicates that the Dockerfile in the current directory should be used."